internal/buffer/buffer.go

```diff
@@ -1167,9 +1167,14 @@ func (b *Buffer) Retab() {
 	tabsize := util.IntOpt(b.Settings["tabsize"])
 	dirty := false
 
-	for i := 0; i < b.LinesNum(); i++ {
-		l := b.LineBytes(i)
+	for i := 0; ; i++ {
+		b.Lock()
+		if i >= b.LinesNum() {
```

Comment:
BTW this commit should be rather before the `highlighter: Fix race ...`, not after it, so that `highlighter: Fix race ...` actually fully fixes the race. Actually, this doesn't need to be a separate commit, it makes sense to just add this locking in `Retab()` already in the commit `buffer: Lock the LineArray in case of modifications ...`, since this is just one of the cases of those LineArray modifications.

Reply:
> This code is actually buggy (completely regardless of our race issue): it bypasses the `EventHandler`, so in particular we cannot undo retab (try it).

Then we should address this with a separate PR, to improve the actual state.